### ISSUE TITLE
Custom yolov5

### ISSUE THREAD

**Issue by davidcg96** [ultralytics/yolov5#8743]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello.
I am doing a project with yolov5 in pytorch, (python).
I want to train the yolov5x network with my own dataset. 
I want to detect people crouching, standing, or walking.
For that I want to load the yolov5x network and modify the output layer to have only 3 classes.
I want to use python in vscode, how can I do that without using jupiter, google collab or kraggle.

I have tried to do this:
`import cv2`
`import torch`
`model = torch.hub.load('ultralytics/yolov5', 'yolov5x',classes=3)`
`model.names.clear()`
`model.names.insert(0,'Depie')`
`model.names.insert(1,'Agachado')`
`model.names.insert(2,'Sentado')`

with this I modify the network for 3 classes and assign the name of the classes?

but...
I want to load my dataset that has this format:
dataset->images->train->walking
dataset->images->train->                                          standing
dataset->images->train->                                          crouching
dataset->images->                               test->walking
dataset->images->     test                                    standing
dataset->images->    test                                     crouching
dataset->images->                               val->walking
 dataset->images->  val                                      standing
 dataset->images->   vak                                     crouching
 dataset->labels->train->walking
 dataset->labels->train                                        standing
 dataset->labels->train                                        crouching
 dataset->labels->                             test->walking
  dataset->labels-> test                                      standing
  dataset->labels->  test                                     crouching
  dataset->labels->                            val->walking
   dataset->labels->val                                      standing
   dataset->labels->val                                      crouching

i use this:
`train_dir = f'D:/python/pruebas/dataset/images/train'`
`label_train_dir=f'D:/python/pruebas/dataset/labels/train' `
`transform = transforms.Compose(
    [transforms.Resize(255),
     transforms.CenterCrop(224),
     transforms.ToTensor()])`
  
`dataset = datasets.ImageFolder(data_dir,transform=transform)`
`dataloader = torch.utils.data.DataLoader(dataset,batch_size=32,shuffle=True)`

this is correct?

To train the model I have to cofigure a .yaml file and pass it to model.train() or with this way of loading data I have to define my own training function?
How can I do this?

**Comment by glenn-jocher** [comment]
@davidcg96 👋 Hello! Thanks for asking about YOLOv5 🚀 **dataset formatting**. To train correctly your data must be in YOLOv5 format. Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths), 2) the number of classes `nc` and 3) a list of class `names`:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes
nc: 80  # number of classes
names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
         'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
         'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
         'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
         'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
         'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
         'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
         'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
         'hair drier', 'toothbrush' ]  # class names
```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<img width="800" alt="Image Labels" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>


Good luck 🍀 and let us know if you have any other questions!

**Comment by davidcg96** [comment]
I have read all the tutorials, and the format of the dataset and its structure is clear to me but the training part is not.
With google collab and looking at the custom dataset tutorial I have managed to train the model.
But I don't want to use google collab because I have a lot of data and google collab closes when I don't use my computer. So I am going to make a python script, which allows me to leave it training for days.

**How do I run with a script.py the file train.py?**
I understand that when using a script, you have to do the following for load the model whit transfer learning and 3 classes:
`model = torch.hub.load('ultralytics/yolov5', 'yolov5x',classes=3)`
`model.names.clear()`
`model.names.insert(0,'Depie')`
`model.names.insert(1,'Agachado')`
`model.names.insert(2,'Sentado')`

How do I train this particular model now? Can you give me an example?

**Comment by glenn-jocher** [comment]
@davidcg96 if you want to modify the names attribute of a PyTorch Hub model you can just do:
```python
model.classes = ['a', 'b', 'c', ...]
```

The training examples are shown in the colab notebook, just run:
```
python train.py --data ... --weights ...
```

**Comment by barney2074** [comment]
@davidcg96 

I think you might be overcomplicating it- someone (Glenn) has done all the hard work for you...!

The first question is: do you want an image classifier (classifies an entire image) or object detection (detects classes of objects within an image)

It sounds like you might want a classifier- which is a bit simpler, and there is a YoloV5 tree and sample data:
refer https://github.com/ultralytics/yolov5/blob/classifier/classifier.py and https://github.com/ultralytics/yolov5/releases/download/v1.0/cifar10.zip
You basically just put images in folders according to class- same as your dataset
<img width="224" alt="Screen Shot 2022-07-28 at 9 47 43 am" src="https://user-images.githubusercontent.com/25944617/181394022-4371481e-3a6c-4fc9-ad49-d31a85bc36ec.png">

If object detection- with YoloV5 it is still fairly easy- you shouldn't need any custom code or to use vscode- you just run train.py
The dataset structure and object classes are defined in the yaml file.
But you will need to annotate all images- so you end up with 2 files for each image- the image itself, plus the label (a text file which defines the object bounding boxes & class ID)

If you are struggling with Colab (you really need the top-tier subscription for any ML work, otherwise it will keep disconnecting mid-job)- the next easiest option is probably running YoloV5 in Docker- it does help if you are familiar with Docker, and have a suitable Linux machine to run it on

Hope this helps

Andrew

**Comment by davidcg96** [comment]
Hi @barney2074, thansk for your answer.
@glenn-jocher 
I have videos from cameras and I want to detect people standing, crouching or sitting. I have to detect people but then I have to analyze in which posture they are. The density of people... I think it's more detection than classification or the same I have to do both.

Regarding the Colab, I don't have the subscription but I'm going to use a Supercomputer with enough hardware for that. This runs a python program, that's why I'm using vscode to do the python program.

Yes, there are two yaml files:
1º Data: this one configures the paths, number of classes etc....
2º Net: this contains how the network is made and the number of classes.

Assuming it is a classifier, I have 3 classes, to do transfer learning and modify the output layer to 3 classes. 
I have to modify the file yolov5x.yaml and put NC=3.
Also modify data.yaml with the routes and classes.
Call train.py

But my doubt, I can do this also in this way:
`model = torch.hub.load('ultralytics/yolov5', 'yolov5x',classes=3)` modify yolov5x.yaml and change nc=3.
`model.train(--epoch --data.yaml)`
or does the model.train function accept only one iterator as dataloader?

**Comment by glenn-jocher** [comment]
@davidcg96 you may use `model.train()` with a single iterator for the dataloader, or pass custom arguments via CLI. For your network, consider modifying the YOLOv5.yaml and data.yaml files, then invoke `model.train()` using a custom `--data` argument. For specific examples see the YOLOv5 documentation, and good luck with your project!
